src/Workspaces/Remote/Core/ExportProviderBuilder.cs

```diff
@@ -0,0 +1,220 @@
+            // This can be important when the MEF catalog records full assembly names such as "System.Runtime, 8.0.0.0" yet
+            // we might be running on .NET 7 or .NET 8, depending on the particular session and user settings.
+            hashContents.Append(Environment.Version.Major);
+
+            foreach (var assemblyPath in assemblyPaths)
+            {
+                // Include assembly path in the hash so that changes to the set of included
+                // assemblies cause the composition to be rebuilt.
+                hashContents.Append(assemblyPath);
+                // Include the last write time in the hash so that newer assemblies written
+                // to the same location cause the composition to be rebuilt.
+                hashContents.Append(File.GetLastWriteTimeUtc(assemblyPath).ToString("F"));
+            }
+
+            // Create base64 string of the hash.
+            var hashAsBase64String = Checksum.Create(hashContents.ToString()).ToString();
+
+            // Convert to filename safe base64 string.
+            return hashAsBase64String.Replace('+', '-').Replace('/', '_').TrimEnd('=');
+        }
+    }
+
+    private static async Task WriteCompositionCacheAsync(string compositionCacheFile, CompositionConfiguration config, bool performCleanup, Action<string> logError, CancellationToken cancellationToken)
+    {
+        try
+        {
+            await Task.Yield().ConfigureAwait(false);
+
+            if (Path.GetDirectoryName(compositionCacheFile) is string directory)
+            {
+                var directoryInfo = Directory.CreateDirectory(directory);
+
+                if (performCleanup)
+                {
+                    // Delete any existing cached files.
+                    foreach (var fileInfo in directoryInfo.EnumerateFiles($"*{CatalogSuffix}"))
+                    {
+                        fileInfo.Delete();
+                        cancellationToken.ThrowIfCancellationRequested();
+                    }
+                }
+            }
+
+            CachedComposition cachedComposition = new();
+            var tempFilePath = Path.Combine(Path.GetTempPath(), Path.GetTempFileName());
+            using (FileStream cacheStream = new(tempFilePath, FileMode.Create, FileAccess.Write, FileShare.None, bufferSize: 4096, useAsync: true))
+            {
+                await cachedComposition.SaveAsync(config, cacheStream, cancellationToken).ConfigureAwait(false);
+            }
+
+#if NET
+            File.Move(tempFilePath, compositionCacheFile, overwrite: true);
+#else
+            // On .NET Framework, File.Move doesn't support overwriting the destination file. Use File.Delete first
+            // to ensure the destination file is removed before moving the temp file. File.Delete will not throw if
+            // the file doesn't exist.
+            File.Delete(compositionCacheFile);
```

Comment:
Why would the destination file exist? Woudln't we have then used that cache in that case?

Reply:
It really shouldn't, will just switch to the File.Move that exists in both TF that doesn't take in the overwrite flag